internal/archiver/archiver.go

```diff
@@ -364,6 +364,30 @@ func (arch *Archiver) Save(ctx context.Context, snPath, target string, previous
 		debug.Log("  %v regular file", target)
 		start := time.Now()
 
+		// check if the file has not changed before performing a fopen operation (more expensive, specially
+		// in network filesystems)
+		if previous != nil && !fileChanged(fi, previous, arch.IgnoreInode) {
```

Comment:
This is copy-pasted from below, right? Why do we need to do it twice?

Reply:
Yes. The PR is work in progress. I pushed early to obtain feedback on it. To refactor that logic one has to take extra care of proper error values inside this function: https://github.com/restic/restic/blob/5c8a3505641300939ac134ae7163006b44b31ab4/internal/archiver/archiver.go#L426

Reply:
> To refactor that logic one has to take extra care of proper error values inside this function: 

I'm not sure I understand what you mean by that. If `fileChanged` is checked before `OpenFile` then there's no reason to check it again afterwards. (Unless we assume that a modified file might be restored to its previous state just between the two previously mentioned methods. And that is very unlikely to happen.)

Reply:
> > To refactor that logic one has to take extra care of proper error values inside this function:
> 
> I'm not sure I understand what you mean by that. If `fileChanged` is checked before `OpenFile` then there's no reason to check it again afterwards. (Unless we assume that a modified file might be restored to its previous state just between the two previously mentioned methods. And that is very unlikely to happen.)


That unlikeness is the reason there is the fopen, to ensure that between those two methods, if a file was modified, you ensure that is still a regular file (a not a symlink).

Reply:
The file is opened primarily because FileSaver.Save needs to read it if it has changed.

The possibility that @MichaelEischer describes is that Lstat in line 347 says regular file, then fileChanged in 369 detects a change, but then fileChanged in line 426 detects that the old file was put back in its place. That is so unlikely that it's hardly worth doing two fileChanged+allBlobsPresent checks for. If the file was restored, FileSaver.Save will detect that by chunking and hashing it.

Reply:
@greatroar in that case I will move the fopen functionality only when the file needs to be read, currently is read for obtaining a newer file info. Let me know if you have other suggestions.

Reply:
`fopen` and the `lstat` afterwards are only necessary when the file should be read again. The additional call to `lstat` is there to ensure that the opened file is still a file. So just remove the second check for `fileChanged`. If the file changes while reading it then there's a race condition anyways and it doesn't matter whether the file info used by `fileChanged` is one millisecond older or newer.

Reply:
I've updated the PR simplifying the logic and removing the second fileChanged check.

Reply:
Please revert the logic simplification. As previously mentioned, that way it is no longer guaranteed that the file info actually matches the file we've opened.

Reply:
@MichaelEischer thanks for the quick review. I applied your suggestion, let me know if there is something else you will like to add.